### PR TITLE
Improve performance when resetting errors.

### LINKED ISSUE #975: Improve performance when resetting errors
The code to reset errors loops through a collection saving each record clearing the  _mailchimp_sync_delta_ and _mailchimp_sync_error_ fields when it should use a single UPDATE query.

This can be remedied by replacing the following code in _resetErrors()_ in **app/code/community/Ebizmarts/MailChimp/Helper/Data.php**
```
       $collection = Mage::getResourceModel('newsletter/subscriber_collection')
            ->addFieldToFilter('mailchimp_sync_error', array('neq' => ''));
        if ($scopeId != 0) {
            $collection = $this->addStoresToFilter($collection, $scopeId, $scope);
        }
        foreach ($collection as $subscriber) {
            $subscriber->setData("mailchimp_sync_delta", '0000-00-00 00:00:00');
            $subscriber->setData("mailchimp_sync_error", '');
            $subscriber->save();
        }
```
with
```
       $storesFilter = '';
        if ($scopeId != 0) {
            $storeIds = $this->getMagentoStoresForMCStoreIdByScope($scopeId, $scope);
            if (count($storeIds) > 0) {
                $storesFilter = array('store_id IN ('.implode(',', $storeIds).')');
            }
        }

        $resource = $this->getCoreResource();
        $conn = $resource->getConnection('core_write');
        $conn->update(
            $resource->getTableName('newsletter/subscriber'),
            array(
                'mailchimp_sync_delta' => '0000-00-00 00:00:00',
                'mailchimp_sync_error' => ''
            ),
            $storesFilter
        );
```

And while we are on the topic of performance there are a couple of instances in the helper where a foreach loop continues even after finding the element being search for.

In _getMagentoStoresForMCStoreIdByScope()_
```
    public function getMagentoStoresForMCStoreIdByScope($scopeId, $scope)
    {
        $ret = array();
        $storeRelation = $this->getStoreRelation();
        $mailchimpStoreIdForScope = $this->getMCStoreId($scopeId, $scope);
        if ($mailchimpStoreIdForScope) {
            foreach ($storeRelation as $mailchimpStoreId => $magentoStoreIds) {
                if ($mailchimpStoreIdForScope == $mailchimpStoreId) {
                    $ret = $magentoStoreIds;
                }
            }
        }
        return $ret;
    }
```
there is no need to even loop for the mailchimp store ID as the array is indexed by mailchimp store ID! That code should be replaced with:
```
    public function getMagentoStoresForMCStoreIdByScope($scopeId, $scope)
    {
        $ret = array();
        $storeRelation = $this->getStoreRelation();
        $mailchimpStoreIdForScope = $this->getMCStoreId($scopeId, $scope);
        if ($mailchimpStoreIdForScope && isset($storeRelation[$mailchimpStoreIdForScope])) {
            $ret = $storeRelation[$mailchimpStoreIdForScope];
        }
        return $ret;
    }
```
Note that _isset()_ is better than _array_key_exists()_ because it is significantly faster and we know that the values can't be NULL.

In _createMergeFields($scopeId, $scope)_ there should be a break after _alreadyExists_ is set to true:
```
                    $alreadyExists = false;
                    $created = false;
                    foreach ($mailchimpFields['merge_fields'] as $mailchimpField) {
                        if ($mailchimpField['tag'] == $chimpTag || strtoupper($chimpTag) == 'EMAIL') {
                            $alreadyExists = true;
                            break; // add this to prevent looping when we have found what we are looking for
                        }
                    }
```

### PR DESCRIPTION
**Pull request by roberto-ebizmarts**
Improving getMagentoStoresForMCStoreIdByScope() method
Extracting code to new method: removeMessageErrorsFromNewsletterSubscriber()
which avoid iterating through a subsribers collection and just does an UPDATE for those records of newsletter_subscriber table which have mailchimp_sync_errors distinct from "". closes #975

### CODE REVIEW

File: app/code/community/Ebizmarts/MailChimp/Helper/Data.php [hunk 1]
```diff
@@ -1101,19 +1100,10 @@ public function getWebhooksKey()
      */
     public function resetErrors($scopeId, $scope = 'stores')
     {
-        // reset subscribers with errors
-        $collection = Mage::getResourceModel('newsletter/subscriber_collection')
-            ->addFieldToFilter('mailchimp_sync_error', array('neq' => ''));
+        $where = "mailchimp_sync_error <> ''";
+        $idsString = null;
```
Comment (Santiagoebizmarts) [inline app/code/community/Ebizmarts/MailChimp/Helper/Data.php]: Rename to storeIdsAsString

File: app/code/community/Ebizmarts/MailChimp/Helper/Data.php [hunk 2]
```diff
@@ -4918,4 +4908,33 @@ protected function getDateHelper()
...
         return Mage::helper('mailchimp/date');
     }
+
+    /**
+     * @param $scopeId
+     * @param $scope
+     * @param $where
+     * @throws Mage_Core_Exception
+     */
+    protected function removeMessageErrorsFromNewsletterSubscriber($scopeId, $scope, $where)
```
Comment (Santiagoebizmarts) [inline app/code/community/Ebizmarts/MailChimp/Helper/Data.php]: Rename to removeErrorsFromSubscribers to maintain the shortest name possible without loosing relevant information.

File: app/code/community/Ebizmarts/MailChimp/Helper/Data.php [hunk 3]
```diff
@@ -1101,19 +1100,10 @@ public function getWebhooksKey()
      */
     public function resetErrors($scopeId, $scope = 'stores')
     {
-        // reset subscribers with errors
-        $collection = Mage::getResourceModel('newsletter/subscriber_collection')
-            ->addFieldToFilter('mailchimp_sync_error', array('neq' => ''));
+        $where = "mailchimp_sync_error <> ''";
```
Comment (Santiagoebizmarts) [inline app/code/community/Ebizmarts/MailChimp/Helper/Data.php]: $where variable should be created inside the removeMessageErrorsFromNewsletterSubscriber function.

File: app/code/community/Ebizmarts/MailChimp/Helper/Data.php [hunk 4]
```diff
@@ -173,17 +173,17 @@ public function getScopeName($scopeArray)
         $storeName = false;
         if (isset($scopeArray['scope'])) {
             switch ($scopeArray['scope']) {
```
[the remark this answers is not in the document]
Comment (roberto-ebizmarts) [reply app/code/community/Ebizmarts/MailChimp/Helper/Data.php]: Solved

File: app/code/community/Ebizmarts/MailChimp/Helper/Data.php (hunk 1 above)
Comment (roberto-ebizmarts) [reply app/code/community/Ebizmarts/MailChimp/Helper/Data.php]: Solved

File: app/code/community/Ebizmarts/MailChimp/Helper/Data.php (hunk 2 above)
Comment (roberto-ebizmarts) [reply app/code/community/Ebizmarts/MailChimp/Helper/Data.php]: Solved

File: app/code/community/Ebizmarts/MailChimp/Helper/Data.php (hunk 3 above)
Comment (roberto-ebizmarts) [reply app/code/community/Ebizmarts/MailChimp/Helper/Data.php]: Solved

**Pull request merged**